### PR TITLE
internal/server: Store URL Token in state to prevent re-registering guest account

### PR DESCRIPTION
**Pull request by briancain**
Prior to this pull request, if a user upgraded or brought back up a Waypoint
server, it would re-register itself with Horizon and get a new URL,
leaving the old guest account behind. This commit fixes that by storing the API
token, and skipping any registering if it exists from prior server
state.

### CODE REVIEW

File: internal/server/singleprocess/service_url.go [hunk 1]
```diff
@@ -11,6 +11,16 @@ import (
 )
 
 func (s *service) initURLGuestAccount(acceptURLTerms bool) error {
+	// Check if API Token already exists, if so, no reason to
+	// re-register and generate a new hostname
+	apiToken, err := s.state.ServerAPITokenGet()
+	if err != nil {
+		return err
+	} else if apiToken != "" {
+		// token already set, guest account already exists
```
Comment (mitchellh) [inline internal/server/singleprocess/service_url.go:20]: I think the success case has to set `s.urlConfig.APIToken` for this to work properly (see L52)

File: internal/server/singleprocess/state/server_apitoken.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package state
+
+import (
+	"github.com/boltdb/bolt"
+)
+
+var (
+	serverAPIToken = []byte("api_token")
```
Comment (mitchellh) [inline internal/server/singleprocess/state/server_apitoken.go:8]: Recommend renaming this and all the function calls to `url_token` and `ServerURLTokenSet` since `api_token` is relatively vague and somewhat confusing cause we also ave API tokens for Waypoint itself and this is unrelated.

**Review by mitchellh — CHANGES_REQUESTED** [review]

File: internal/server/singleprocess/service_url.go [hunk 3]
```diff
@@ -40,5 +50,9 @@ func (s *service) initURLGuestAccount(acceptURLTerms bool) error {
 	}
 
 	s.urlConfig.APIToken = accountResp.Token
+	if err := s.state.ServerURLTokenSet(accountResp.Token); err != nil {
```
Comment (briancain) [inline internal/server/singleprocess/service_url.go:53]: Responding to your comment about setting the token in the success case...I should be doing it here for the initial check at the beginning of this function to work... is that what you meant @mitchellh ?

Comment (mitchellh) [reply internal/server/singleprocess/service_url.go:53]: I don't think so.

I think you just need to put:

> s.urlConfig.APIToken = accountResp.Token

where my original comment is. I think all this here is fine.

Comment (briancain) [reply internal/server/singleprocess/service_url.go:53]: Oh right, I see what you mean now! :sweat_smile: :+1:

**Review by evanphx — APPROVED** [review]

**Review by mitchellh — APPROVED** [review]
